modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/FlowControlledQueueMessageDeliveryImpl.java

```diff
@@ -60,38 +61,46 @@ public int deliverMessageToSubscriptions(StorageQueue storageQueue) throws Andes
 
         int numOfConsumers = currentSubscriptions.size();
         int consumerIndexCounter = 0;
+        /*
+         *This is the minimum cursor that is the smallest cursor from all the bound subscriptions for a queue.
+         * All messages before this cursor is removed from the buffer.
+         */
+        long minCursor = Long.MAX_VALUE;
+        boolean isRemoved = false;
 
-        while (iterator.hasNext()) {
-            try {
+        try {
 
-                DeliverableAndesMetadata message = iterator.next();
-                int numOfCurrentMsgDeliverySchedules = 0;
-                boolean subscriberWithMatchingSelectorFound = true;
-                boolean suspendedSubFound = false;
+            int numOfCurrentMsgDeliverySchedules = 0;
 
                 /*
                  * if message is addressed to queues, only ONE subscriber should
                  * get the message. Otherwise, loop for every subscriber
                  */
-                for (int j = 0; j < numOfConsumers; j++) {
-                    int currentConsumerIndex = consumerIndexCounter % numOfConsumers;
-                    AndesSubscription localSubscription = currentSubscriptions.get(currentConsumerIndex);
-                    consumerIndexCounter = consumerIndexCounter + 1;
-                    if (localSubscription.getSubscriberConnection().isSuspended()) {
-                        suspendedSubFound = true;
-                        continue;
-                    }
-                    if (localSubscription.getSubscriberConnection().hasRoomToAcceptMessages()) {
-
+            for (int j = 0; j < numOfConsumers; j++) {
+                int currentConsumerIndex = consumerIndexCounter % numOfConsumers;
+                AndesSubscription localSubscription = currentSubscriptions.get(currentConsumerIndex);
+                long currentCursor = localSubscription.getCursor();
+                consumerIndexCounter = consumerIndexCounter + 1;
+                if (localSubscription.getSubscriberConnection().isSuspended()) {
+                    continue;
+                }
+                Iterator<Map.Entry<Long, DeliverableAndesMetadata>> iterator = messages.tailMap(currentCursor)
+                        .entrySet().iterator();
+                while (iterator.hasNext()) {
+                    Map.Entry<Long, DeliverableAndesMetadata> entry = iterator.next();
+                    DeliverableAndesMetadata message = entry.getValue();
+                    currentCursor = entry.getKey();
+                    if (checkIfUnacked(currentSubscriptions, currentCursor)) {
+                        iterator.remove();
+                        isRemoved = true;
+                    } else if (localSubscription.getSubscriberConnection().hasRoomToAcceptMessages()) {
                         if (!localSubscription.getSubscriberConnection().
                                 isMessageAcceptedByConnectionSelector(message)) {
```

Comment:
It becomes longer if we do so and formatting breaks the line at message. I guess this is better than breaking the line at message